internal/server/tokens_test.go

```diff
@@ -0,0 +1,202 @@
+package server
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gotest.tools/v3/assert"
+
+	"github.com/infrahq/infra/internal/access"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/internal/server/providers"
+)
+
+func TestAPI_CreateToken(t *testing.T) {
+	srv := setupServer(t, withAdminUser)
+	routes := srv.GenerateRoutes(prometheus.NewRegistry())
+
+	_, err := data.InitializeSettings(srv.db)
+	assert.NilError(t, err)
+
+	type testCase struct {
+		setup    func(t *testing.T, req *http.Request)
+		expected func(t *testing.T, resp *httptest.ResponseRecorder)
+	}
+
+	run := func(t *testing.T, tc testCase) {
+		req, err := http.NewRequest(http.MethodPost, "/api/tokens", nil)
+		assert.NilError(t, err)
+		req.Header.Add("Infra-Version", apiVersionLatest)
+
+		if tc.setup != nil {
+			tc.setup(t, req)
+		}
+
+		resp := httptest.NewRecorder()
+		routes.ServeHTTP(resp, req)
+
+		tc.expected(t, resp)
+	}
+
+	testCases := map[string]testCase{
+		"not authenticated": {
+			expected: func(t *testing.T, resp *httptest.ResponseRecorder) {
+				assert.Equal(t, resp.Code, http.StatusUnauthorized)
+			},
+		},
+		"infra provider user with valid access key": {
+			setup: func(t *testing.T, req *http.Request) {
+				user := &models.Identity{
+					Name: "spike@example.com",
+				}
+				err := data.CreateIdentity(srv.db, user)
+				assert.NilError(t, err)
+				_, err = data.CreateProviderUser(srv.db, data.InfraProvider(srv.db), user)
+				assert.NilError(t, err)
+
+				key := &models.AccessKey{
+					IssuedFor:  user.ID,
+					ProviderID: data.InfraProvider(srv.db).ID,
+					ExpiresAt:  time.Now().Add(10 * time.Second),
+				}
+				accessKey, err := data.CreateAccessKey(srv.db, key)
+				assert.NilError(t, err)
+
+				req.Header.Set("Authorization", "Bearer "+accessKey)
+			},
+			expected: func(t *testing.T, resp *httptest.ResponseRecorder) {
+				assert.NilError(t, err)
+				assert.Equal(t, resp.Code, http.StatusCreated)
```

Comment:
A check for the contents of `resp.Body` would be great too in one of these cases. Not blocking though, this is great additional test coverage.